Fixes #4329 "Back button in edit categories triggers back of media details."

4D17Y4 opened the pull request:
**Description (required)**

Fixes #4329

What changes did you make and why?

Added onKey listener to the fragment view.

**Tests performed (required)**

Tested BetaDebug on Nokia 6.1+ with API level 29.

4D17Y4 commented:
Up for review @nicolas-raoul , @neslihanturan .

neslihanturan reviewed (changes requested):
Hey @4D17Y4 , thanks the logic seems okay to me, however I frequently get this error when I am testing this PR. Do you think if this is related? 
```
USER_COMMENT=
APP_VERSION_CODE=876
APP_VERSION_NAME=3.0.0-debug-4D17Y4-backbutton
ANDROID_VERSION=10
PHONE_MODEL=Redmi Note 8
STACK_TRACE=java.lang.IllegalStateException: The application's PagerAdapter changed the adapter's contents without calling PagerAdapter#notifyDataSetChanged! Expected adapter item count: 240, found: 260 Pager id: fr.free.nrw.commons.beta:id/mediaDetailsPager Pager class: class androidx.viewpager.widget.ViewPager Problematic adapter: class fr.free.nrw.commons.media.MediaDetailPagerFragment$MediaDetailAdapter
	at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1143)
	at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1092)
	at androidx.viewpager.widget.ViewPager.onMeasure(ViewPager.java:1622)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6872)
	at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
	at android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
	at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6872)
	at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6872)
	at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
	at android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
	at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6872)
	at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
	at android.view.View.measure(View.java:25086)
	at android.widget.RelativeLayout.measureChildHorizontal(RelativeLayout.java:735)
	at android.widget.RelativeLayout.onMeasure(RelativeLayout.java:481)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6872)
	at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
	at android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
	at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6872)
	at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
	at androidx.appcompat.widget.ContentFrameLayout.onMeasure(ContentFrameLayout.java:143)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6872)
	at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
	at android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
	at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6872)
	at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6872)
	at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1552)
	at android.widget.LinearLayout.measureVertical(LinearLayout.java:842)
	at android.widget.LinearLayout.onMeasure(LinearLayout.java:721)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6872)
	at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
	at com.android.internal.policy.DecorView.onMeasure(DecorView.java:742)
	at android.view.View.measure(View.java:25086)
	at android.view.ViewRootImpl.performMeasure(ViewRootImpl.java:3083)
	at android.view.ViewRootImpl.measureHierarchy(ViewRootImpl.java:1857)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2146)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1745)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7768)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:967)
	at android.view.Choreographer.doCallbacks(Choreographer.java:791)
	at android.view.Choreographer.doFrame(Choreographer.java:726)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:952)
	at android.os.Handler.handleCallback(Handler.java:883)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7356)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:940)

IS_SILENT=false
```

4D17Y4 commented:
I think its another issue #4283
I'll check and confirm if this commit has something to do with it.

4D17Y4 commented:
The mentioned issue is resolved.
This should not create a problem now.

nicolas-raoul reviewed (commented):
It fixes the issue in the case:
Tap picture, Tap pencil, Tap Back

I wonder if fixing this case would also be possible? (I know it is not described in the issue but very close) :
Tap picture, Tap pencil, Tap search icon, Tap Back (keyboard disappears), Tap Back (now back to contributions instead of back to media details)

4D17Y4 commented:
Removed the focus logic as it needed much modification to track every event on the fragment and was not the best way.


Used other logic to check if the mediaDetails is closed or open.
* Added a parameter and function to store the present fragment as we need to access it to get the mediaDetails state.
* Used this instance to check if media details is open or not. 
* Using the functions return value identified when to close the fragment.


I've modified every instance of the Pager fragment to keep it consistent on each mediaDetails.
This handles each and every case and also the case @nicolas-raoul  pointed out.

Need confirmation on the function name.

Let me know if I missed something.

Up for review @nicolas-raoul , @neslihanturan ,
Thanks

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:
```diff
@@ -385,6 +392,9 @@ public void onDataSetChanged() {
     //FragmentStatePagerAdapter allows user to swipe across collection of images (no. of images undetermined)
     private class MediaDetailAdapter extends FragmentStatePagerAdapter {
 
+        // Keeps track of the current displayed fragment.
```
Would you mind using the `/**` javadoc notation? Thanks!

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:
```diff
@@ -414,5 +424,30 @@ public int getCount() {
         }
+
+        /**
+         * getter for mCurrentFragment
```
-> `Get the currently displayed fragment.`
Javadoc is written once (usually) and read by many developers (and shown by IDEs), so it is worth the effort :-)

4D17Y4 replied:
done, Thanks

nicolas-raoul reviewed (changes requested):
I tested the changes on many activities, it works great!
The source code looks good too, I only request very minors changes.

neslihanturan reviewed (approved):
Works great, we can merge it as soon as @nicolas-raoul 's reviews are done :)

4D17Y4 commented:
Done @neslihanturan, Thanks.

neslihanturan commented:
Merging as @nicolas-raoul 's reviews are done, thanks @4D17Y4 :)
